File: iosApp/iosApp/Presentation/Main/CafeManagement/CafeManagementUiState.swift
//
//  CafeManagementUiState.swift
//  ConCafe
//
//  Created by 홍희표 on 2026/03/07.
//

import Foundation
import Shared

struct CafeManagementUiState {
    var ownedCafes: [CafeManagementData.OwnedCafeSummary] = []
    var searchableCafes: [CafeManagementData.SearchableCafeSummary] = []
    var pendingClaims: [CafeManagementData.PendingClaimSummary] = []
    var isLoading = true
    var isShowingAllCafes = false
    var cafeSearchQuery = ""
    var infoMessage: String?

    var featuredCafe: CafeManagementData.OwnedCafeSummary? {
        ownedCafes.first
    }

    var hasOwnedCafes: Bool {
        !ownedCafes.isEmpty
    }

    var visibleOwnedCafes: [CafeManagementData.OwnedCafeSummary] {
        isShowingAllCafes ? ownedCafes : Array(ownedCafes.prefix(Self.defaultVisibleCafeCount))
    }

    var hasHiddenOwnedCafes: Bool {
        ownedCafes.count > Self.defaultVisibleCafeCount
    }

    var filteredSearchableCafes: [CafeManagementData.SearchableCafeSummary] {
        searchableCafes.filter { cafe in
            cafeSearchQuery.isEmpty ||
            cafe.name.localizedCaseInsensitiveContains(cafeSearchQuery) ||
            cafe.location.localizedCaseInsensitiveContains(cafeSearchQuery)
        }
    }

    private static let defaultVisibleCafeCount = 2
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/presentation/main/cafemanagement/CafeManagementUiState.kt
package com.hhp227.concafe.presentation.main.cafemanagement

import com.hhp227.concafe.domain.model.CafeManagementData

data class CafeManagementUiState(
    val ownedCafes: List<CafeManagementData.OwnedCafeSummary> = emptyList(),
    val searchableCafes: List<CafeManagementData.SearchableCafeSummary> = emptyList(),
    val pendingClaims: List<CafeManagementData.PendingClaimSummary> = emptyList(),
    val isShowingAllCafes: Boolean = false,
    val cafeSearchQuery: String = "",
    val infoMessage: String? = null
) {
    val featuredCafe: CafeManagementData.OwnedCafeSummary?
        get() = ownedCafes.firstOrNull()

    val hasOwnedCafes: Boolean
        get() = ownedCafes.isNotEmpty()

    val visibleOwnedCafes: List<CafeManagementData.OwnedCafeSummary>
        get() = if (isShowingAllCafes) ownedCafes else ownedCafes.take(DEFAULT_VISIBLE_CAFE_COUNT)

    val hasHiddenOwnedCafes: Boolean
        get() = ownedCafes.size > DEFAULT_VISIBLE_CAFE_COUNT

    val filteredSearchableCafes: List<CafeManagementData.SearchableCafeSummary>
        get() = searchableCafes.filter {
            cafeSearchQuery.isBlank() ||
                it.name.contains(cafeSearchQuery, ignoreCase = true) ||
                it.location.contains(cafeSearchQuery, ignoreCase = true)
        }

    companion object {
        private const val DEFAULT_VISIBLE_CAFE_COUNT = 2
    }
}
